apiserver/internal/handlers/resource/resource.go

```diff
@@ -0,0 +1,408 @@
+	api "github.com/juju/juju/api/client/resources"
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/core/logger"
+	coreresource "github.com/juju/juju/core/resource"
+	"github.com/juju/juju/domain/resource"
+	resourceerrors "github.com/juju/juju/domain/resource/errors"
+	charmresource "github.com/juju/juju/internal/charm/resource"
+	"github.com/juju/juju/internal/errors"
+	"github.com/juju/juju/rpc/params"
+)
+
+// Validator validates downloaded resource blobs.
+type Validator interface {
+	Validate(
+		reader io.ReadCloser,
+		expectedSHA384 string,
+		expectedSize int64,
+	) (io.ReadCloser, error)
+}
+
+// ResourceHandler is the HTTP handler for client downloads and
+// uploads of resources.
+type ResourceHandler struct {
+	logger                logger.Logger
+	authFunc              func(*http.Request, ...string) (names.Tag, error)
+	changeAllowedFunc     func(context.Context) error
+	resourceServiceGetter ResourceServiceGetter
+	validator             Validator
+}
+
+// NewResourceHandler returns a new HTTP client resource handler.
+func NewResourceHandler(
+	logger logger.Logger,
+	authFunc func(*http.Request, ...string) (names.Tag, error),
+	changeAllowedFunc func(context.Context) error,
+	resourceServiceGetter ResourceServiceGetter,
+	validator Validator,
+) *ResourceHandler {
+	return &ResourceHandler{
+		logger:                logger,
+		authFunc:              authFunc,
+		changeAllowedFunc:     changeAllowedFunc,
+		resourceServiceGetter: resourceServiceGetter,
+		validator:             validator,
+	}
+}
+
+// ServeHTTP implements http.Handler.
+func (h *ResourceHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	tag, err := h.authFunc(req, names.UserTagKind, names.MachineTagKind, names.ControllerAgentTagKind, names.ApplicationTagKind)
+	if err != nil {
+		if err := sendError(resp, err, h.logger); err != nil {
+			h.logger.Errorf("%v", err)
+		}
+		return
+	}
+
+	resourceService, err := h.resourceServiceGetter.Resource(req)
+	if err != nil {
+		if err := sendError(resp, err, h.logger); err != nil {
+			h.logger.Errorf("returning error to user: %v", err)
+		}
+		return
+	}
+
+	switch req.Method {
+	case "GET":
+		reader, size, err := h.download(resourceService, req)
+		if err != nil {
+			if err := sendError(resp, err, h.logger); err != nil {
+				h.logger.Errorf("%v", err)
+			}
+			return
+		}
+		defer reader.Close()
+		header := resp.Header()
+		header.Set("Content-Type", params.ContentTypeRaw)
+		header.Set("Content-Length", fmt.Sprint(size))
+		resp.WriteHeader(http.StatusOK)
+		var bytesWritten int64
+		if bytesWritten, err = io.Copy(resp, reader); err != nil {
+			if err := sendError(resp, errors.Errorf("resource download failed: %w", err), h.logger); err != nil {
+				h.logger.Errorf("returning error to user: %v", err)
+			}
+			return
+		}
+		if bytesWritten != size {
+			h.logger.Errorf("resource download size does not match expected resource size: %d != %d", bytesWritten, size)
+		}
```

Comment:
This shouldn't be an error, probably a warning.

Errors mean, an operator of Juju has to take some immediate action about this. Otherwise, everything is an error and it's just noise.